Sources/RInAppMessaging/Router.swift

```diff
@@ -124,11 +142,90 @@ internal class Router: RouterType {
+                image: UIImage(data: imageBlob)!,
+                backgroundColor: UIColor(hexString: tooltipData.backgroundColor) ?? .white))
+
+            let onClose = {
+                self.displayedTooltips[identifier]?.removeFromSuperview()
+                self.displayedTooltips[identifier] = nil
+                completion()
+            }
+
+            tooltipView.onImageTap = {
+                if let uriToOpen = URL(string: tooltipData.bodyData.redirectURL ?? "") {
+                    UIApplication.shared.open(uriToOpen)
+                    onClose()
+                }
+            }
+            tooltipView.onExitButtonTap = onClose
+
+            let superview = self.findParentViewForTooltip(targetView)
+            guard superview != targetView else {
+                Logger.debug("Cannot find suitable view for tooltip targeting: \(targetView.description)")
+                return
+            }
+
+            superview.addSubview(tooltipView)
+            superview.layoutIfNeeded()
+            if let displayedCampaign = superview.findIAMViewSubview() {
+                superview.bringSubviewToFront(displayedCampaign)
+            }
+
+            self.updateFrame(targetView: targetView, tooltipView: tooltipView, superview: superview, position: position)
+
+            if let parentScrollView = superview as? UIScrollView {
+                let observer = parentScrollView.observe(\.contentOffset, options: [.new, .old]) { [weak self] layer, change in
+                    self?.updateFrame(targetView: targetView, tooltipView: tooltipView, superview: superview, position: position)
+                    if self?.isTooltipVisible(tooltipView) == true {
+                        becameVisibleHandler(tooltipView)
+                    }
+                }
+                self.observers.append(WeakWrapper(value: observer))
+            } else {
+                let observer = targetView.observe(\.frame, options: [.new, .old]) { [weak self] layer, change in
+                    self?.updateFrame(targetView: targetView, tooltipView: tooltipView, superview: superview, position: position)
+                    if self?.isTooltipVisible(tooltipView) == true {
+                        becameVisibleHandler(tooltipView)
+                    }
+                }
+                self.observers.append(WeakWrapper(value: observer))
+            }
+
+            // TOOLTIP: Keep tooltip away from sceen edges
+    //        let distFromRightEdge = view.window!.bounds.maxX - tooltipView.frame.maxX
+    //        if distFromRightEdge < UIConstants.minDistFromEdge {
+    //            tooltipView.frame.origin.x -= UIConstants.minDistFromEdge - distFromRightEdge
+    //        }
+    //
+    //        let distFromLeftEdge = UIConstants.minDistFromEdge - tooltipView.frame.minX
+    //        if distFromLeftEdge < UIConstants.minDistFromEdge {
+    //            tooltipView.frame.origin.x += distFromLeftEdge
+    //        }
```

Comment:
Is the above needed?

Reply:
Future improvement. For now it can be deleted